### PR TITLE
TOF workflow

### PR DESCRIPTION
**Pull request by noferini**
inserting TOF workflow:
Clustering moved from digitizer_workflow to tof workflow
track matching and calibration task in workflow
caching if TOF matrixes + refinmenets

### CODE REVIEW

**Comment by noferini** [comment]
@chiarazampolli @shahor02 
Old pull request with Ruben's comment
https://github.com/AliceO2Group/AliceO2/pull/2207

**Comment by shahor02** [comment]
Hi @noferini ,
I still see some backup and link files, 
https://github.com/AliceO2Group/AliceO2/pull/2214/files#diff-14
https://github.com/AliceO2Group/AliceO2/pull/2214/files#diff-23

**Comment by noferini** [comment]
> Hi @noferini ,
> I still see some backup and link files,
> https://github.com/AliceO2Group/AliceO2/pull/2214/files#diff-14
> https://github.com/AliceO2Group/AliceO2/pull/2214/files#diff-23

The backup files are a my mistake I will fix it.
About the links: it seems to me that git flagged with "->" the move from a directory to another directory, but they are not links. Am I wrong?

**Comment by shahor02** [comment]
@noferini probably yes, but the second link leads to https://github.com/AliceO2Group/AliceO2/blob/bc92664a44ba2e2d80dbef490d05c96ad2f9a56a/Detectors/GlobalTrackingWorkflow/tofworkflow/src/%23TOFClusterWriterSpec.cxx%23, now you have committed files ``#TOFClusterWriterSpec.cxx#`` and ``TOFClusterWriterSpec.cxx``

**Comment by noferini** [comment]
Hi Ruben,

we implemented most of your comments.

- We removed one backup file

- We removed a couple of files (we don't need them):

DataFormats/Detectors/TOF/include/DataFormatsTOF/CompressedFormat.h

DataFormats/Detectors/TOF/include/DataFormatsTOF/Decoder.h

- We fixed data member names in our structs

- cable lengths moved to const (but not constexpr because we inizialize the values in a cxx file, because it is a very big matrix)

- default constructor used as suggested in Encoder class

- data member pointer initialized to null values

We didn't implement this suggestion:

- TOF DPL: suggestion "better to convert to task" all our Specs

We prefer to do that in the next pull request because functionality checks (more time) are needed to guarantee that results are correct.

One more question:
Do you still see problems with backup/link files?

**Comment by noferini** [comment]
For your information, some examples on how to run our DPL:
_________________________________________________________
o2-tof-reco-workflow
input: TOF digits
output: TOF cluster, TOF matching info, TOF calib info
_________________________________________________________
o2-tof-reco-workflow --output-type clusters
input: TOF digits
output: TOF cluster
_________________________________________________________
o2-tof-reco-workflow --input-type clusters
input: TOF cluster
output: TOF matching info, TOF calib info
_________________________________________________________
o2-tof-reco-workflow --input-type clusters --output-type matching-info
input: TOF cluster
output: TOF matching info

File: Detectors/GlobalTracking/src/MatchTOF.cxx [hunk 1]
```diff
@@ -665,29 +765,34 @@ void MatchTOF::doMatching(int sec)
...
           LOG(DEBUG) << "MATCHING FOUND: We have a match! between track " << mTracksSectIndexCache[indices[0]][itrk] << " and TOF cluster " << mTOFClusSectIndexCache[indices[0]][itof];
           foundCluster = true;
           evIdx eventIndexTOFCluster(trefTOF.getEntryInTree(), mTOFClusSectIndexCache[indices[0]][itof]);
           evIdx eventIndexTracks(mCurrTracksTreeEntry, mTracksSectIndexCache[indices[0]][itrk]);
-          mMatchedTracksPairs.emplace_back(std::make_pair(eventIndexTracks, o2::dataformats::MatchInfoTOF(eventIndexTOFCluster, chi2, trkLTInt[iPropagation]))); // TODO: check if this is correct!
+          mMatchedTracksPairs.emplace_back(o2::dataformats::MatchInfoTOF(eventIndexTOFCluster, chi2, trkLTInt[iPropagation], eventIndexTracks)); // TODO: check if this is correct!
+
+#ifdef _ALLOW_TOF_DEBUG_
           for (int ilabel = 0; ilabel < labelsTOF.size(); ilabel++) {
             LOG(DEBUG) << "TOF label " << ilabel << ": trackID = " << labelsTOF[ilabel].getTrackID() << ", eventID = " << labelsTOF[ilabel].getEventID() << ", sourceID = " << labelsTOF[ilabel].getSourceID();
```
Comment (shahor02) [inline Detectors/GlobalTracking/src/MatchTOF.cxx]: note that the MCCompLabel has changed: the fake is not signaled by label<0 anymore, but by the special flag isFake, see #2171
You better use directly the label streaming method, then the complete info will be printed:
 ``LOG(DEBUG) << "TOF label " << ilabel << labelsTOF[ilabel] ;``

File: Detectors/GlobalTracking/src/MatchTOF.cxx [hunk 2]
```diff
@@ -702,32 +807,37 @@ void MatchTOF::selectBestMatches()
...
+                               int(mTOFClusWork[matchingPair.getTOFClIndex()].getTimeRaw() * 1E12), // add time stamp
+                               mTOFClusWork[matchingPair.getTOFClIndex()].getTimeRaw() - matchingPair.getLTIntegralOut().getTOF(o2::track::PID::Pion),
+                               mTOFClusWork[matchingPair.getTOFClIndex()].getTot());
+
+    const auto& labelTPC = (*mTPCLabels)[matchingPair.getTrackIndex()];
+    const auto& labelITS = (*mITSLabels)[matchingPair.getTrackIndex()];
+    // we want to store positive labels independently of how they are flagged from TPC,ITS people
+    //    o2::MCCompLabel labelTPC(abs(labelTPCor.getTrackID()), labelTPCor.getEventID(), labelTPCor.getSourceID());
+    //    o2::MCCompLabel labelITS(abs(labelITSor.getTrackID()), labelITSor.getEventID(), labelITSor.getSourceID());
     LOG(DEBUG) << "labelTPC: trackID = " << labelTPC.getTrackID() << ", eventID = " << labelTPC.getEventID() << ", sourceID = " << labelTPC.getSourceID();
```
Comment (shahor02) [inline Detectors/GlobalTracking/src/MatchTOF.cxx]: idem

File: DataFormats/Detectors/TOF/include/DataFormatsTOF/Encoder.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+  bool open(std::string name);
+  bool alloc(long size);
+  bool encode(/*define input structure*/);
+  bool flush();
+  bool close();
+  void setVerbose(bool val) { mVerbose = val; };
+
+  // benchmarks
+  double mIntegratedBytes = 0.;
```
[the remark this answers is not in the document]
Comment (noferini) [reply DataFormats/Detectors/TOF/include/DataFormatsTOF/Encoder.h]: now private

File: DataFormats/Detectors/TOF/src/Encoder.cxx [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+  mUnion->CrateHeader.mustBeOne = 1;
+  //    mUnion->CrateHeader.drmID = summary.DRMGlobalHeader.drmID;
+  //    mUnion->CrateHeader.eventCounter = summary.DRMGlobalTrailer.LocalEventCounter;
+  //    mUnion->CrateHeader.bunchID = summary.DRMStatusHeader3.l0BCID;
+#ifdef VERBOSE
+  if (mVerbose) {
+    auto BunchID = mUnion->CrateHeader.bunchID;
+    auto EventCounter = mUnion->CrateHeader.eventCounter;
+    auto DRMID = mUnion->CrateHeader.drmID;
+    std::cout << boost::format("%08x") % mUnion->Data
```
[the remark this answers is not in the document]
Comment (noferini) [reply DataFormats/Detectors/TOF/src/Encoder.cxx]: For the moment cout line commented and boost header no longer included

File: DataFormats/Detectors/TOF/src/Encoder.cxx [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+    /** check if TRM is empty **/
+    //      if (summary.nTRMSpiderHits[itrm] == 0)
+    //	continue;
+
+    unsigned char firstFilledFrame = 255;
+    unsigned char lastFilledFrame = 0;
+
+    /** loop over hits **/
+    /*
```
[the remark this answers is not in the document]
Comment (noferini) [reply DataFormats/Detectors/TOF/src/Encoder.cxx]: It has to be filled. I prefer to keep the comment for the moment

File: DataFormats/Reconstruction/include/ReconstructionDataFormats/CalibInfoTOF.h [hunk 6]
```diff
@@ -14,6 +14,8 @@
 #ifndef ALICEO2_CALIBINFOTOF_H
```
[the remark this answers is not in the document]
Comment (noferini) [reply DataFormats/Reconstruction/include/ReconstructionDataFormats/CalibInfoTOF.h]: I moved all standalone tof classes in tof dir

File: Detectors/GlobalTracking/include/GlobalTracking/MatchTOF.h [hunk 7]
```diff
@@ -158,7 +167,13 @@ class MatchTOF
   void fillTOFmatchTree(const char* tname, int cacheTOF, int sectTOF, int plateTOF, int stripTOF, int padXTOF, int padZTOF, int cacheeTrk, int crossedStrip, int sectPropagation, int platePropagation, int stripPropagation, int padXPropagation, int padZPropagation, float resX, float resZ, float res, o2::dataformats::TrackTPCITS& trk, float intLength, float intTimePion, float timeTOF);
```
[the remark this answers is not in the document]
Comment (noferini) [reply Detectors/GlobalTracking/include/GlobalTracking/MatchTOF.h]: These come from the past. We will discuss how to clean this part of the code

File: Detectors/GlobalTracking/src/MatchTOF.cxx (hunk 1 above)
Comment (noferini) [reply Detectors/GlobalTracking/src/MatchTOF.cxx]: LOG fixed. About the new fake flag I have to digest the change

File: Detectors/GlobalTracking/src/MatchTOF.cxx (hunk 2 above)
Comment (noferini) [reply Detectors/GlobalTracking/src/MatchTOF.cxx]: done

**Comment by noferini** [comment]
> Hi @noferini
> some comments from the previous review were not addressed, I've copied them here, see below (also added new ones).
> There still stray files:
> shahoian@alicers02:~/alice/O2$ find . -name "*.*" | grep '#'
> ./Detectors/GlobalTrackingWorkflow/tofworkflow/src/#TOFClusterWriterSpec.cxx#
Removed


> ./Detectors/GlobalTrackingWorkflow/tofworkflow/include/TOFWorkflow/#ClusterReaderSpec.h#
I didn't find this file (???)


> 
> Ok for the conversion to tasks, but what about moving the DPL devices to appropriate directories (#2207 (comment)) ?
> The GlobalTrackingWorkflow is for the specs and workflows which require (have dependency on) multiple detectors. All single detector devices, like TOFClusterizerSpec, ClusterReaderSpec, DigitReaderSpec etc. would belong to Detectors/TOF/workflow/ and referred by e.g. tof-reco-workflow.cxx in the Detectors/GlobalTrackingWorkflow (no need to put it in separate directory GlobalTrackingWorkflow/tofworkflow). Also for your tofworkflow/src/TrackReaderSpec.cxx the natural place would be where the ITSTPC matching is done and it should be renamed to e.g. TrackTPCITSReaderSpec.

Almost done. I need only to remove TrackReaderSpec by replacing it with the TrackTPCITSReaderSpec.

File: Detectors/GlobalTrackingWorkflow/tofworkflow/src/ClusterReaderSpec.cxx [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+    LOG(INFO) << "TOF: Sending ROMode= " << roMode << " to GRPUpdater";
+    pc.outputs().snapshot(Output{ "TOF", "ROMode", 0, Lifetime::Timeframe }, roMode);
+  } else {
+    LOG(ERROR) << "Cannot read the TOF clusters !";
+    return;
+  }
+
+  mState = 2;
+  //  pc.services().get<ControlService>().readyToQuit(false);
```
[the remark this answers is not in the document]
Comment (noferini) [reply Detectors/GlobalTrackingWorkflow/tofworkflow/src/ClusterReaderSpec.cxx]: done

File: Detectors/GlobalTrackingWorkflow/tofworkflow/src/DigitReaderSpec.cxx [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+    LOG(INFO) << "TOF: Sending ROMode= " << roMode << " to GRPUpdater";
+    pc.outputs().snapshot(Output{ "TOF", "ROMode", 0, Lifetime::Timeframe }, roMode);
+  } else {
+    LOG(ERROR) << "Cannot read the TOF digits !";
+    return;
+  }
+
+  mState = 2;
+  //  pc.services().get<ControlService>().readyToQuit(false);
```
[the remark this answers is not in the document]
Comment (noferini) [reply Detectors/GlobalTrackingWorkflow/tofworkflow/src/DigitReaderSpec.cxx]: done

**Comment by noferini** [comment]
TrackReaderSpec replaced by TrackTPCITSReaderSpec.
Therefore, I think I covered all points.
Let us know if you think that something is still missing.
Cheers

**Comment by chiarazampolli** [comment]
Hi Ruben, 
Yes, I am doing it incrementally, sorry, because we had also other comments that Francesco addressed yesterday. 
I will finish today.
Cheers,
Chiara

**Comment by chiarazampolli** [comment]
Hello @noferini , @shahor02 ,

I committed some fixes, and open the PR on Francesco's branch.
One question for @shahor02 (or someone else, I don't know): I find the output of the o2checkcode a bit difficult to read (I hope I got all errors), isn't there a way to make it more explicit?

Chiara

**Comment by noferini** [comment]
PR accepted

**Comment by shahor02** [comment]
Hi @chiarazampolli , the remaining checkode errors are not related to this PR, they come from the Framework. I think this PR can be merged, but I cannot do this if some of the tests are red, @ktf can.

**Comment by ktf** [comment]
These is the usual clang bug. I need to remember how to mask it (and maybe we should bump clang for the code checker to a version which does not have the problem, now...).

**Comment by ktf** [comment]
I assume you are happy with the 23 commits for this PR?

**Comment by shahor02** [comment]
Indeed, better to squash them, but doing this at merging will create misleading commit name, @noferini, could you do this manually?

**Comment by noferini** [comment]
I'll squash commits this afternoon.
Cheers

**Comment by noferini** [comment]
I'll squash commits this afternoon.
Cheers

**Comment by noferini** [comment]
I did the squash (now 4 commits).
No diff with respect previous status.
cheers

**Comment by shahor02** [comment]
@noferini Thanks, due to the o2checkcode still failing for unrelated reasons I cannot merge it, @ktf, could you please do this?

**Comment by ktf** [comment]
Yes, let me just wait a bit more for the O2/o2 test.

**Pull request merged**